internal/state/state.go

```diff
@@ -1,50 +1,81 @@
 package state
 
 import (
-	"math/big"
-	"relayer/bindings/BridgeBase"
+	"relayer/internal/types"
+	"sort"
 
-	log "github.com/sirupsen/logrus"
-
-	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
 )
 
-// BlockchainClient defines the interface for blockchain operations required by getStateWithProof.
-type BlockchainClient interface {
-	LatestBlockNumber() (*big.Int, error)                                                  // Returns the block number
-	GetStateWithProof(opts *bind.CallOpts) (BridgeBase.SharedStructsStateWithProof, error) // Gets the state with proof
-	FinalizationInterval() *big.Int                                                        // Gets the finalization interval
+type State struct {
+	stakes      map[common.Address]int32
+	latestEpoch uint64
+	totalStake  int32
+	stakeGetter func(common.Address) int32
+}
+
+func NewState(getter func(common.Address) int32) *State {
+	return &State{
+		stakes:      make(map[common.Address]int32),
+		stakeGetter: getter,
+	}
 }
 
-// getStateWithProof fetches the state with proof for a given epoch and block number using the provided client.
-func GetStateWithProof(client BlockchainClient, logger *log.Logger, epoch *big.Int, block_num *big.Int) (*BridgeBase.SharedStructsStateWithProof, error) {
-	if block_num == nil {
-		block, err := client.LatestBlockNumber()
-		if err != nil || block == nil {
-			logger.WithField("block", block).WithError(err).Fatal("BlockByNumber")
+func (s *State) UpdateState(block *types.PillarBlock) {
+	if s.latestEpoch > uint64(block.Epoch) {
+		return
+	}
```

Comment:
It is possible to have a pillarBlock with the same epoch. But we still need to apply vote count changes from it